src/pykeen/models/baseline/models.py

```diff
@@ -0,0 +1,184 @@
+# -*- coding: utf-8 -*-
+
+"""Non-parametric baseline models."""
+
+from abc import ABC
+from typing import Optional
+
+import numpy
+import scipy.sparse
+import torch
+
+from .utils import get_relation_similarity, marginal_score, get_csr_matrix
+from ..base import Model
+from ...triples import CoreTriplesFactory
+
+__all__ = [
+    'EvaluationOnlyModel',
+    'MarginalDistributionBaseline',
+    'SoftInverseTripleBaseline',
+]
+
+
+class EvaluationOnlyModel(Model, ABC):
+    """A model which only implements the methods used for evaluation."""
+
+    def __init__(self, triples_factory: CoreTriplesFactory):
+        """Non-parametric models take a minimal set of arguments.
+
+        :param triples_factory: The training triples factory is used to assign the number of entities, relations,
+            and inverse condition in the non-parametric model.
+        """
+        super().__init__(
+            triples_factory=triples_factory,
+            # These operations are deterministic and a random seed can be fixed
+            # just to avoid warnings
+            random_seed=0,
+            # These operations do not need to be performed on a GPU
+            preferred_device='cpu',
+        )
+
+    def _reset_parameters_(self):
+        """Non-parametric models do not implement :meth:`Model._reset_parameters_`."""
+        raise RuntimeError
+
+    def collect_regularization_term(self):  # noqa:D102
+        """Non-parametric models do not implement :meth:`Model.collect_regularization_term`."""
+        raise RuntimeError
+
+    def score_hrt(self, hrt_batch: torch.LongTensor):  # noqa:D102
+        """Non-parametric models do not implement :meth:`Model.score_hrt`."""
+        raise RuntimeError
+
+    def score_r(self, ht_batch: torch.LongTensor):  # noqa:D102
+        """Non-parametric models do not implement :meth:`Model.score_r`."""
+        raise RuntimeError
+
+
+class MarginalDistributionBaseline(EvaluationOnlyModel):
+    r"""
+    Score based on marginal distributions.
+
+    To predict scores for the tails, we simplify
+
+    .. math ::
+        P(t | h, r) = P(t | h) * P(t | r)
+
+    Depending on the settings, we either set $P(t | *) = \frac{1}{n}$, or estimate them by counting occurrences in the
+    training triples.
+
+    .. note ::
+        This model cannot make use of GPU acceleration, since internally it uses scipy's sparse matrices.
+
+    ---
+    name: Marginal Distribution Baseline
+    citation:
+        author: Berrendorf
+        year: 2021
+        link: https://github.com/pykeen/pykeen/pull/514
+        github: pykeen/pykeen
+    """
+
+    def __init__(
+        self,
+        triples_factory: CoreTriplesFactory,
+        entity_margin: bool = True,
```

Comment:
what happens if `entity_margin` and `relation_margin` are both false? Should that raise an exception?

Reply:
I first has this exception, but then reverted it again. It will lead to a uniform distribution, i.e. equal scores for all entities.

Reply:
Updated text in 1f1b977e